Notebook clustering

shruthi-raghuraman opened the pull request:
Notebook contains cleaning and clustering. Clustering still has to be improved so that prediction can be better.

shruthi-raghuraman committed:
Notebook to commit

shruthi-raghuraman committed:
Notebook for clustering

fridex commented:
Nice work! Few comments after a brief review:

I'm not sure about the fixed index to the error messages, but based on samples present in the notebook, it looks like it works.

What I would expect out of clusters:

Random tokens made for the path to temporary directories should not occur (see for example cluster 99) - they should be unique for installation - as they show up in the output, the clustering is probably not that precise. It might be a result of tf-idf that assigns high importance to tokens that are unique for documents, but in this case, we are more interested in tokens shared across documents that subsequently make clusters.

I see the following words in multiple clusters: "print", "parentheses", "missing", "did", "mean", call", "syntaxerror'. That's correctly classified error for packages that fail to install due to legacy print statement from Python2:

```
$ python3 -c "print 'foo'"                                                   
  File "<string>", line 1
    print 'foo'
              ^
SyntaxError: Missing parentheses in call to 'print'. Did you mean print('foo')?
``` 

I would expect this to be ideally one cluster of error types.

It feels like this is getting better output than the original naive clustering. Do you have any ideas on how to improve even more?

BTW please do not parse solvers using `solver_id[0:19]`, we will soon change the suffix (and the "core" part may vary based on solver name) , you can do `solver_document_id.rsplit("-", maxsplit=1)[0]`

shruthi-raghuraman committed:
Heirarchical Clustering implemented

shruthi-raghuraman commented:
Thanks for the comments Frido!
Have now taken out the fixed error and added get_message_component function that gets value with identifier.
The clusters make more sense: for ex: Cluster 4 is the specific syntax error.

Output of clusters of data are now in dataframes. So instead of tokens for kmeans you use a linkage between two documents. So you have the error of sum of squares as the optimized objective function (Ward linkage) as the criterion to create a cluster. I still use tf-idf but perhaps there are other ways to get weights?

Also fixed the parsing solver using fixed index.

fridex commented:
> Thanks for the comments Frido!
> Have now taken out the fixed error and added get_message_component function that gets value with identifier.
> The clusters make more sense: for ex: Cluster 4 is the specific syntax error.
> 
> Output of clusters of data are now in dataframes. So instead of tokens for kmeans you use a linkage between two documents. So you have the error of sum of squares as the optimized objective function (Ward linkage) as the criterion to create a cluster. I still use tf-idf but perhaps there are other ways to get weights?
> 
> Also fixed the parsing solver using fixed index.

Awesome, thanks! :+1: 

Very nice presentation and work :100: :1st_place_medal:
